### PR TITLE
Fixed #34483 -- Fixed timesince()/timeuntil() with timezone-aware dates and interval less than 1 day.

### PR DESCRIPTION
**Pull request by nessita**
Regression in 8d67e16493c903adc9d049141028bc0fff43f8c8.

Thanks Lorenzo Peña for the report.

### CODE REVIEW

**Comment by nessita** [comment]
The core of the issue is located in this piece of code:

```python
    total_months = (now.year - d.year) * 12 + (now.month - d.month)
    if d.day > now.day or (d.day == now.day and d.time() > now.time()):
        total_months -= 1
    years, months = divmod(total_months, 12)
```

At first, the simpler fix seemed to be ensuring that we decrement `total_months` only if non-zero, but after a broader look, I think the correct fix is the one proposed so we can ensure that any existing and future calculations using `now` (at least when `now` is not given) operates over the same timezone.

File: docs/releases/4.2.1.txt [hunk 1]
```diff
@@ -33,3 +33,7 @@ Bugfixes
 * Fixed a regression in Django 4.2 where creating copies and deep copies of
   ``HttpRequest``, ``HttpResponse``, and their subclasses didn't always work
   correctly (:ticket:`34482`, :ticket:`34484`).
+
+* Fixed a regression in Django 4.2 where ``timesince()`` would return incorrect
```
Comment (felixxm) [inline docs/releases/4.2.1.txt:37]: `timeuntil` is also affected:

_"Fixed a regression in Django 4.2 where ``timesince`` and ``timeuntil`` template filters would ..."_

**Review by felixxm — COMMENTED** [review]
@nessita Thanks :+1: I left comments. 


Also, a commit message should be in [the past tense](https://docs.djangoproject.com/en/4.2/internals/contributing/committing-code/#committing-guidelines) and describes what was fixed, e.g.

_"Fixed #34483 -- Fixed timesince()/timeuntil() with timezone-aware dates and interval less than 1 day."_

Please don't use backticks in commit messages as they are unnecessary and only interpreted on GitHub, and add "#" before the ticket number.

**Comment by felixxm** [comment]
> The core of the issue is located in this piece of code:
> 
> ```python
>     total_months = (now.year - d.year) * 12 + (now.month - d.month)
>     if d.day > now.day or (d.day == now.day and d.time() > now.time()):
>         total_months -= 1
>     years, months = divmod(total_months, 12)
> ```
> 
> At first, the simpler fix seemed to be ensuring that we decrement `total_months` only if non-zero, but after a broader look, I think the correct fix is the one proposed so we can ensure that any existing and future calculations using `now` (at least when `now` is not given) operates over the same timezone.

Agreed, the incorrect use of `.time()` is crucial. What do you think about using `delta` instead? This solves the issue without changing the `now` timezone (which hasn't been changed in 8d67e16493c903adc9d049141028bc0fff43f8c8):

```diff
diff --git a/django/utils/timesince.py b/django/utils/timesince.py
index 690556751a..94ba24d48a 100644
--- a/django/utils/timesince.py
+++ b/django/utils/timesince.py
@@ -63,11 +63,7 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
     if now and not isinstance(now, datetime.datetime):
         now = datetime.datetime(now.year, now.month, now.day)
 
-    # Ensure that the default `now` uses the same timezone as the given `d`
-    # datetime object. This way, future operations on both `d` and `now` will
-    # always use the same timezone. Specifically, the incoming comparison
-    # `d.time() > now.time()` should operate on exactly the same timezone.
-    now = now or datetime.datetime.now(tz=d.tzinfo if is_aware(d) else None)
+    now = now or datetime.datetime.now(datetime.timezone.utc if is_aware(d) else None)
 
     if reversed:
         d, now = now, d
@@ -81,7 +77,8 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
 
     # Get years and months.
     total_months = (now.year - d.year) * 12 + (now.month - d.month)
-    if d.day > now.day or (d.day == now.day and d.time() > now.time()):
+    time_delta = delta - datetime.timedelta(days=delta.days)
+    if d.day > now.day or (d.day == now.day and time_delta.total_seconds() < 0):
         total_months -= 1
     years, months = divmod(total_months, 12)
 
```

File: docs/releases/4.2.1.txt [hunk 2]
```diff
@@ -33,3 +33,7 @@ Bugfixes
...
+  results for a datetime using a non-UTC timezone and whose time difference
+  with the current time is within the day (:ticket:`34483`).
```
Comment (felixxm) [inline docs/releases/4.2.1.txt:39]: Small rewording suggestion:
```suggestion
* Fixed a regression in Django 4.2 where ``timesince`` and ``timeuntil``
  template filters returned incorrect results for a datetime with a non-UTC
  timezone when a time difference is less than 1 day (:ticket:`34483`).
```

Comment (nessita) [reply docs/releases/4.2.1.txt:39]: Applied.

File: tests/utils_tests/test_timesince.py [hunk 3]
```diff
@@ -241,6 +242,13 @@ def test_depth_invalid(self):
         with self.assertRaisesMessage(ValueError, msg):
             timesince(self.t, self.t, depth=0)
 
+    def test_a_second_ago_with_zoneinfo(self):
```
[the remark this answers is not in the document]
Comment (nessita) [reply tests/utils_tests/test_timesince.py:245]: I think that the adding of this decorator is not necessary: on one hand, there is already a second test case that would run all these test but with the tz support enabled (`TZAwareTimesinceTests`), on the other hand, this issue is present and needs fixing also when tz support is not enabled so having tests covering both cases is necessary.

File: tests/utils_tests/test_timesince.py [hunk 4]
```diff
@@ -241,6 +242,13 @@ def test_depth_invalid(self):
...
+        """Timesince should not return negative year values (#34483)."""
```
[the remark this answers is not in the document]
Comment (nessita) [reply tests/utils_tests/test_timesince.py:246]: I considered it an important piece of the code so future readers understand why the test is there. This follows the existing pattern for other tests, like `test_date_objects` and `test_both_date_objects`.

Is there a more recent guideline to not add individual docstrings to test methods?

File: tests/utils_tests/test_timesince.py [hunk 5]
```diff
@@ -241,6 +242,13 @@ def test_depth_invalid(self):
...
+        now_in_berlin = timezone.now().astimezone(
+            zoneinfo.ZoneInfo(key="Europe/Berlin")
+        )
+        self.assertEqual(timesince(now_in_berlin - self.onesecond), "0\xa0minutes")
```
[the remark this answers is not in the document]
Comment (nessita) [reply tests/utils_tests/test_timesince.py:250]: Will do, thanks!

File: tests/utils_tests/test_timesince.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply tests/utils_tests/test_timesince.py:245]: This decorator is not about tz support in Django but in OS.

File: tests/utils_tests/test_timesince.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply tests/utils_tests/test_timesince.py:246]: A lot of all tests use old patterns, the current policy for Django is not to change them (old docstrings) just for consistency.

> I considered it an important piece of the code so future readers understand why the test is there.

About the ticket number: _"Reserve ticket references for obscure issues where the ticket has additional details that can’t be easily described in docstrings or comments"_ which is not the case here.

I still don't see any value in the docstring. Assertions are self-explanatory and the sentence _"should not return negative year values"_ doesn't add any details that can be useful in the future, it's not important that it returned negative values, it's important that it didn't returned correct values and in this case assertions are enough to see what is expected.

File: tests/utils_tests/test_timesince.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply tests/utils_tests/test_timesince.py:245]: > the other test case already have it on (both). What's your rationale to add it in this testcase/test?

I don't understand :thinking: Only one test in `TimesinceTests` use real time zones i.e. `test_different_timezones` and it's decorated by `requires_tz_support`. What do you mean by _"the other test case already have it on (both)"_?

[the remark this answers is not in the document]
Comment (nessita) [reply tests/utils_tests/test_timesince.py]: I understand your point now, sorry for dragging this across 4 comments. Adding the decorator now!

**Comment by felixxm** [comment]
@nessita Thanks for updates :+1: Please squash commits and update the commit message.

**Comment by nessita** [comment]
I was trying to reproduce the reported failure:
```
======================================================================
FAIL [4.965s]: test_middleware_caching (template_tests.test_response.CacheMiddlewareTest.test_middleware_caching)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-pg-server-side-binding/database/postgres/label/focal-pr/python/python3.11/tests/template_tests/test_response.py", line 407, in test_middleware_caching
    self.assertEqual(response.content, response2.content)
AssertionError: b'This is where you can find the snark: /snark/\n1681389259.685652\n' != b'This is where you can find the snark: /snark/\n1681389264.634695\n'

----------------------------------------------------------------------
Ran 16259 tests in 671.253s

FAILED (failures=1, skipped=283, expected failures=8)
```
using a focal installation with python 3.11 and pointing the test suite to a PG database. Is this a test that fail sometimes? Shall I ignore it for now?

**Comment by felixxm** [comment]
> using a focal installation with python 3.11 and pointing the test suite to a PG database. Is this a test that fail sometimes? Shall I ignore it for now?

I don't see such failure :thinking:

**Comment by nessita** [comment]
> @nessita Thanks for updates +1 Please squash commits and update the commit message.

Done!

**Comment by nessita** [comment]
> > using a focal installation with python 3.11 and pointing the test suite to a PG database. Is this a test that fail sometimes? Shall I ignore it for now?
> 
> I don't see such failure thinking

It was run by Jenkins, reporting failure until a few minutes ago. I don't understand! I assumed it was now green because you or someone else somehow re-run it?

https://djangoci.com/job/pull-requests-pg-server-side-binding/database=postgres,label=focal-pr,python=python3.11/997/

**Comment by felixxm** [comment]
> > @nessita Thanks for updates +1 Please squash commits and update the commit message.
> 
> Done!

Thanks :+1: Welcome aboard :boat:

**Comment by felixxm** [comment]
> > > using a focal installation with python 3.11 and pointing the test suite to a PG database. Is this a test that fail sometimes? Shall I ignore it for now?
> > 
> > 
> > I don't see such failure thinking
> 
> It was run by Jenkins, reporting failure until a few minutes ago. I don't understand! I assumed it is now green because you or someone else somehow re-run it?
> 
> https://djangoci.com/job/pull-requests-pg-server-side-binding/database=postgres,label=focal-pr,python=python3.11/997/

Nope, you can ignore it.

**Comment by felixxm** [comment]
@nessita Unfortunately, there are 2 issues with this patch :disappointed: First, Django 4.2 supports Python 3.8 so we should have changed the import when backporting, see #16763. Second, it seems that there is still an issue when a time difference is less than 1 day but days are different, see [logs](https://djangoci.com/job/django-main/2424/).

**Comment by nessita** [comment]
> @nessita Unfortunately, there are 2 issues with this patch disappointed First, Django 4.2 supports Python 3.8 so we should have changed the import when backporting, see #16763. Second, it seems that there is still an issue when a time difference is less than 1 day but days are different, see [logs](https://djangoci.com/job/django-main/2424/).

Will work on this ASAP!

**Pull request merged**